src/librustc_mir/interpret/operator.rs

```diff
@@ -197,94 +137,129 @@ impl<'a, 'mir, 'tcx, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
-            (Ge, _) => PrimVal::from_bool(l >= r),
+        if left_ty.is_signed() {
+            let op: Option<fn(&i128, &i128) -> bool> = match bin_op {
+                Lt => Some(i128::lt),
+                Le => Some(i128::le),
+                Gt => Some(i128::gt),
+                Ge => Some(i128::ge),
+                _ => None,
+            };
+            if let Some(op) = op {
+                let l = self.sign_extend(l, left_ty)? as i128;
+                let r = self.sign_extend(r, right_ty)? as i128;
+                return Ok((PrimVal::from_bool(op(&l, &r)), false));
+            }
+            let op: Option<fn(i128, i128) -> (i128, bool)> = match bin_op {
+                Rem | Div if r == 0 => return Ok((PrimVal::Bytes(l), true)),
+                Div => Some(i128::overflowing_div),
+                Rem => Some(i128::overflowing_rem),
+                Add => Some(i128::overflowing_add),
+                Sub => Some(i128::overflowing_sub),
+                Mul => Some(i128::overflowing_mul),
+                _ => None,
+            };
+            if let Some(op) = op {
+                let l128 = self.sign_extend(l, left_ty)? as i128;
+                let r = self.sign_extend(r, right_ty)? as i128;
+                let size = self.layout_of(left_ty)?.size.bits();
+                match bin_op {
+                    Rem | Div => {
+                        // int_min / -1
+                        if r == -1 && l == (1 << (size - 1)) {
+                            return Ok((PrimVal::Bytes(l), true));
+                        }
+                    },
+                    _ => {},
+                }
+                trace!("{}, {}, {}", l, l128, r);
+                let (result, mut oflo) = op(l128, r);
+                trace!("{}, {}", result, oflo);
+                if !oflo && size != 128 {
+                    let max = 1 << (size - 1);
+                    oflo = result >= max || result < -max;
+                }
+                let result = result as u128;
+                let truncated = self.truncate(result, left_ty)?;
+                return Ok((PrimVal::Bytes(truncated), oflo));
+            }
+        }
 
-            (BitOr, _) => PrimVal::Bytes(l | r),
-            (BitAnd, _) => PrimVal::Bytes(l & r),
-            (BitXor, _) => PrimVal::Bytes(l ^ r),
+        if let ty::TyFloat(fty) = left_ty.sty {
+            return Ok((float_op(bin_op, l, r, fty), false));
+        }
 
-            (Add, k) if k.is_int() => return int_arithmetic!(k, overflowing_add, l, r),
-            (Sub, k) if k.is_int() => return int_arithmetic!(k, overflowing_sub, l, r),
-            (Mul, k) if k.is_int() => return int_arithmetic!(k, overflowing_mul, l, r),
-            (Div, k) if k.is_int() => return int_arithmetic!(k, overflowing_div, l, r),
-            (Rem, k) if k.is_int() => return int_arithmetic!(k, overflowing_rem, l, r),
+        // only ints left
```

Comment:
Does this mean to say "unsigned integers"?

Reply:
no, all ops that don't differ between signed and unsigned are handled here, too